bio-demultiplexing-fasta/src/demultiplexed.py

```diff
@@ -35,34 +36,111 @@ def importMetadata(metadata: CustomSample, outputDir: Path) -> Path:
+    for sample in samples:
+        if sample.name.startswith("_metadata"):
+            continue
+
+        for filePath in sample.path.glob("*.fastq.gz"):
+            filePath.link_to(outFolder / filePath.name)
+
+
+def loadSingleEnd(sample: CustomSample) -> Tuple[Path, str]:
+    sample.unzip()
+
+    filePathList = list(sample.path.glob("*.fastq*"))
+    if len(filePathList) == 1:
+        filePath = filePathList[0]
+        return filePath, filePath.name.split("_")[0]
+
+    raise ValueError(f">> [Microbiome analysis] Sample \"{sample.name}\" must contain exactly one fastq file")
+
+
+def loadPairedEnd(sample: CustomSample) -> Tuple[Path, Path, str]:
+    sample.unzip()
+
+    forwardPathList = list(sample.path.glob("*_R1_*.fastq*"))
+    reversePathList = list(sample.path.glob("*_R2_*.fastq*"))
+
+    if len(forwardPathList) != 1 or len(reversePathList) != 1:
+        raise ValueError(f">> [Microbiome analysis] Invalid paired-end sample: {sample.name}. Must contain 2 files, one with \"_R1_\" and another with \"_R2_\" in name")
+
+    forwardPath = forwardPathList[0]
+    reversePath = reversePathList[0]
+
+    return forwardPath, reversePath, forwardPath.name.split("_")[0]
+
+
+def isGzCompressed(dataset: CustomDataset) -> bool:
+    for sample in dataset.samples:
+        sample.unzip()
+        if len(list(sample.path.glob("*.fastq.gz"))) > 0:
+            return True
+
+    return False
+
+
 def importDemultiplexedSamples(
-    fastqSamples: List[CustomSample],
-    metadata: CustomSample,
+    dataset: CustomDataset,
     experiment: Experiment,
     outputDataset: CustomDataset,
     outputDir: Path
 ):
 
-    manifestPath = outputDir / "manifest.tsv"
-    createManifest(fastqSamples, manifestPath)
-
-    demuxZipPath = importSample(manifestPath, experiment.parameters["sequenceType"], outputDir)
+    logging.info(">> [Microbiome analysis] Preparing data for import into QIIME2 format")
+    metadata = ctx_qiime2.getMetadataSample(experiment.dataset)
+
+    pairedEnd = isPairedEnd(dataset)
+    if isGzCompressed(dataset):
+        inputPath = outputDir / "fastqGzFolder"
+        inputPath.mkdir()
+        dumpGzFiles(dataset.samples, inputPath)
+        inputFormat = "CasavaOneEightSingleLanePerSampleDirFmt"
+
+        if pairedEnd:
+            sequenceType = "SampleData[PairedEndSequencesWithQuality]"
+        else:
+            sequenceType = "SampleData[SequencesWithQuality]"
+    else:
+        inputPath = outputDir / "manifest.tsv"
+        fastqSamples = ctx_qiime2.getFastqDPSamples(dataset)
+
+        if pairedEnd:
+            createManifestPaired(fastqSamples, inputPath)
+            sequenceType = "SampleData[PairedEndSequencesWithQuality]"
+            inputFormat = "PairedEndFastqManifestPhred33V2"
+        else:
+            createManifestSingle(fastqSamples, inputPath)
+            sequenceType = "SampleData[SequencesWithQuality]"
+            inputFormat = "SingleEndFastqManifestPhred33V2"
```

Comment:
Let's extract all this logic into coretexpylib `SequenceDataset` and `SequenceSample` classes. `SequenceSample` currently supports `[fa, fq, fasta, fastq]` files with or without gz compression. Additional support for `_metadata` and `paired-end` sequences has to be implemented.

Reply:
Yes, this is a very common functionality and will be used across large amount of data jobs in the future. It definitely belongs in coretexpylib.